internal/app/synchronizer/synchronizer_service.go

```diff
@@ -46,6 +132,165 @@ func (s *synchronizerService) Register(ctx context.Context, req *pb.RegisterRequ
 // added to the list of proposals to be evaluated in the current cycle. At the
 //  end of the cycle, the user defined evaluation method is triggered and the
 // matches accepted by it are returned as results.
-func (s *synchronizerService) EvaluateProposals(ctx context.Context, req *pb.EvaluateProposalsRequest) (*pb.EvaluateProposalsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+func (s *synchronizerService) EvaluateProposals(ctx context.Context, req *ipb.EvaluateProposalsRequest) (*ipb.EvaluateProposalsResponse, error) {
+	synchronizerServiceLogger.Debugf("Request to evaluate propsals - id:%v, proposals: %v", req.Id, getMatchIds(req.Match))
+	// pendingRequests keeps track of number of requests pending. This is incremented
+	// in addProposals while holding the state mutex. The count should be decremented
+	// only after this request completes.
+	err := s.addProposals(req.Id, req.Match)
+	if err != nil {
+		return nil, err
+	}
+
+	// If proposals were successfully added, then the pending request count has been
+	// incremented and should only be reduced when this request is completed.
+	defer s.cycleData.pendingRequests.Done()
```

Comment:
Also because this is crossing method boundaries this may break WaitGroup. Not sure though but there's weird behavior. Make sure there's sufficient test coverage to vet this.

Reply:
Not sure if I understand yufan's concern here. (Add proposal adds to the wait group only when it is past the failure point). After this, if EvaluateProposals fails at any point, the deferred pendingRequests.Done() will reduce this count correctly. The idea is, once there is work queued, we set the defer s.cycleData.pendingRequests.Done() so whenever the method terminates in any way, it will mark the queued work as done. 

@jeremyje - yes, that was a concern. I validated that it indeed works (the wg variable is stored on the service object and is not crossing message boundaries). I ve tests validating that new requests correctly queue behind a set of requests from previous cycle.

Reply:
Ah but if add proposals encountered an error, evaluate proposals will return directly at line 142 so that line 147 won’t be executed

Reply:
Ah but if add proposals encountered an error, evaluate proposals will return directly at line 142 so that line 147 won’t be executed

Reply:
Yes :) and add proposal only calls Add() after all errors return paths are done with. It would have been nice and readable to get the Add right before this defer out of add proposal - but that is a race. We need to hold the state mutex and do the increment to the wg and the actual addition of proposals together. Hence what we do is - wait for all error conditions to be done - then increment the wg and append the list.

Reply:
A quick fix is to move line 147 to line 141

Reply:
I had it there and the -race caught a race with it :). I can explain better f2f when you are in but moving to 141 introduces a race yielding incorrectness. Further, it could actually set a defer done before we have really added any proposals and then you have a problem if add fails.  I don't yet see the actual issue in the current logic
that needs to be fixed. We can discuss f2f when you are in.